smac/intensification/successive_halving.py

```diff
@@ -0,0 +1,440 @@
+                cutoff = self._adapt_cutoff(challenger=challenger,
+                                            incumbent=incumbent,
+                                            run_history=run_history,
+                                            inc_sum_cost=inc_sum_cost)
+                if cutoff is not None and cutoff <= 0:
+                    # no time to validate challenger
+                    self.logger.debug("Stop challenger itensification due to adaptive capping.")
+                    break
+
+                # setting cutoff based on the type of budget & adaptive capping
+                tae_cutoff = min(cutoff, budget) if self.budget_type == 'cutoff' else cutoff
+
+                self.logger.debug('Cutoff for challenger: %s' % str(cutoff))
+
+                # run target algorithm for each instance-seed pair
+                self.logger.debug("Execute target algorithm")
+                try:
+                    status, cost, dur, res = self.tae_runner.start(
+                        config=challenger,
+                        instance=instance,
+                        seed=seed,
+                        cutoff=tae_cutoff,
+                        instance_specific=self.instance_specifics.get(instance, "0"),
+                        capped=(self.cutoff is not None) and
+                               (cutoff < self.cutoff)
+                    )
+                    self._ta_time += dur
+                    self._num_run += 1
+
+                except CappedRunException:
+                    # We move on to the next configuration if we reach maximum cutoff i.e., capped
+                    self.logger.debug("Budget exhausted by adaptive capping; "
+                                      "Interrupting current challenger and moving on to the next one")
+                    capped_configs.append(challenger)
+                    break
+
+            # count every challenger exactly once per SH run
+            if first_run:
+                self._chall_indx += 1
+
+        # eliminate capped configuration from the race & reset capped_configs
+        _ = [challengers.remove(c) for c in capped_configs]
+
+        return challengers
+
+    def _top_k(self, configs: typing.List[Configuration],
+               run_history: RunHistory,
+               k: int):
+        """
+        selects the top 'k' configurations from the given list based on their performance in this budget
+
+        Parameters
+        ----------
+        configs: typing.List[Configuration]
+            list of configurations to filter from
+        run_history: RunHistory
+            stores all runs we ran so far
+        k: int
+            number of configurations to select
+
+        Returns
+        -------
+        challengers: typing.List[Configuration]
+            top challenger configurations, sorted in increasing costs
+        """
+        # NOTE run_history overwrites the cost for runs since it doesnt record each budget (cutoff) separately
```

Comment:
I wonder whether this will hurt us quite a bit when we will combine HB with BO. Nothing to do for now, but we have to keep that in mind and have to change the runhistory at some point

Reply:
But in this section, we have to use `sum_cost` to compute the incumbent runtime for adaptive capping. The incumbent cost is calculated the same way in intensification.py too.

https://github.com/automl/SMAC3/blob/6f77eafaafe36d5c59c88689bfaea5d9505c3003/smac/intensification/intensification.py#L342-L348